### PR TITLE
add executable and c opts

### PR DESCRIPTION
**Pull request by nathantchan**
Go compile command now written to executable file to avoid arg list being too long. Also, add ability for c compile options to be passed to _cgo_codegen function. @bsilver8192 @AustinSchuh #42

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by nathantchan** [comment]
I signed it!

**Comment by yugui** [comment]
@nathantchan Why do we need to have different sets of options for code generation and actual compilation?

**Comment by brian-peloton** [comment]
One major limitation is that the actual compilation takes place via `cc_library`, which has very different escaping rules than the existing copts attr of `_cgo_codegn_rule`. In particular, `cc_library` does make variable and location expansion which means while `_cgo_codegn_rule` sets some environment variables as alternatives. I can't think of any way to create a single string which indicates a path relative to the root of the execdir for both of them, and it seems like adding make variable or location expansion to the Go rules will also add a lot of unnecessary complication.

Also, even with allow location expansion, I don't see any way to get directory names, which are important for things like `-I` flags.

**Comment by yugui** [comment]
@brian-peloton 
I am trying to make `_cgo_codegen_rule` as consistent to `cc_library` as possible. As you said, make variable and location expansion are what are missing in the current implementation.  So it is a bug we can fix.

Could you show me concrete use cases about other directory things?

@nathantchan I'd like to discuss this topic about C options separately.  Could you exclude that part from this PR?

File: go/def.bzl [hunk 1]
```diff
@@ -153,11 +161,13 @@ def emit_go_asm_action(ctx, source, out_obj):
       " ".join(args),
   ]
 
+  f = _tmp_cmd_file(cmds, ctx, out_obj + "GoAsmCompileFile")
```
Comment (yugui) [inline go/def.bzl]: Could you name the temp file `out_obj + ".GoAsmCompile.params"` to be consistent to follow convention in Bazel?

File: go/def.bzl [hunk 2]
```diff
@@ -133,6 +133,14 @@ def go_environment_vars(ctx):
                                    {"GOOS": "linux",
                                     "GOARCH": "amd64"})
 
+def _tmp_cmd_file(cmds, ctx, fn):
```
Comment (yugui) [inline go/def.bzl]: Can we call this function `emit_generate_params_action` or something?
I prefer the suffix `_action` to be consistent to other actions, and I prefer the name `params` to follow a convention in Bazel.

**Comment by nathantchan** [comment]
I was not quite sure how to incorporate all the changes suggested into this PR, so I created a new one: https://github.com/bazelbuild/rules_go/pull/87

**Comment by yugui** [comment]
Can I close this PR for now because #87 was merged?

**Pull request closed without merging**